ddtrace/internal/datadog/profiling/dd_wrapper/src/interface.cpp

```diff
@@ -1,33 +1,39 @@
-// Unless explicitly stated otherwise all files in this repository are licensed
-// under the Apache License Version 2.0. This product includes software
-// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
-// Datadog, Inc.
-
 #include "interface.hpp"
-#include "exporter.hpp"
+#include "global_cache.hpp"
+#include "libdatadog_helpers.hpp"
+#include "profile.hpp"
+#include "sample.hpp"
+#include "uploader.hpp"
 
 #include <cstdlib>
 #include <iostream>
 #include <unistd.h>
 
 // State
-bool is_initialized = false;
-Datadog::Uploader* g_uploader;
-Datadog::Profile* g_profile;
-Datadog::Profile* g_profile_real[2];
+bool is_ddup_initialized = false;
 bool g_prof_flag = true;
 
-// State used only for one-time configuration
-Datadog::UploaderBuilder uploader_builder;
-Datadog::ProfileBuilder profile_builder;
+// Store a pointer to the current profile.  This can be assumed set in inner
+// functions, but any function that may be called by a new thread should check
+thread_local Datadog::Sample* g_profile;
+
+// When a fork is detected, we need to reinitialize this state.
+// This handler will be called in the single thread of the child process after the fork
+void
+ddup_fork_handler()
+{
+    Datadog::Profile::mark_dirty();
```

Comment:
don't we also need a barrier for the uploader?

Reply:
Ah, indeed.  I had originally thought of doing this with a barrier, but I changed it to an exclusive lock to protect new uploads from sneaking in during the `fork()` operation.